### ISSUE TITLE
Calling secured Webapi from secured Webapi including User credentials

### ISSUE THREAD

**Issue by steffler412** [IdentityServer/IdentityServer3#3847]
Hello,

we use IdentityServer3 for Authentication and Authorization in our company. RessourceOwner, Implicit and Hybrid Flows are present in our client applications. Behind the scenes there is a complex architecture including secured webapis using UseIdentityServerBearerTokenAuthentication. 

But I am getting confused at the point where an secured API calls another secured API and the second API needs the user information who called the first API with his client. All I can find is that you recommend the client credentials flow in this case, but in this flow is no user included.
Is there any recommended way to reuse the accesstoken in API1 to call API2? There is an "ugly" way to get to accesstoken from the Requestheader, but this solution seems not to be the way you should do it...

In other words I want to request an accesstoken including scopes for api1 and api2 to the user´s client. Then request api1 to fetch some data. In api1 is some businesslogic that needs to call api2 with the user´s identity included.

Am I getting something wrong here?

**Comment by brockallen** [comment]
You want an extension grant: https://identityserver4.readthedocs.io/en/release/topics/extension_grants.html

**Comment by steffler412** [comment]
Thanks for the quick response.
Do you recommend the implementation for an extension grant rather than requesting an access token with all relevant api scopes included and reusing the token on api side? Last solution seems like the simplest way...

**Comment by brockallen** [comment]
That requires your client to know internal implementation details of your API. That's leaky.

**Comment by steffler412** [comment]
Seems legit.
I have implemented the extension grant like above for simple delegation to exchange the token to another one with the api2 scope. 
Everything works fine, but this custom grant flow lasts about 600 to 1000ms for every request from api1 to api2. That duration is too long, especially there´s an api3,4,5,6... like a microservice architecture. 

Do you recommend to cache exchanged long lived tokens on api side? Is there an existing .NET library for that? 
Or is there a different quick solution, to get the user´s identity and the calling clientId from the token on the backend apis?

**Comment by brockallen** [comment]
Yes, caching is the solution.
